themes/default/content/blog/safe-lambda-secrets/index.md

```diff
@@ -0,0 +1,545 @@
+{{% /notes %}}
+
+## Structured configuration
+
+We're going to start by adding settings, both secret and plaintext, to our [stack configuration]({{< relref "/docs/intro/concepts/config" >}}), in a [structured way]({{< relref "/docs/intro/concepts/config#structured-configuration" >}}). We're going to add these settings as a dictionary so we can use the names as the variable names in our Lambda Function.
+
+First we add two settings in plaintext that are going to be used as environment variables in our lambda:
+
+```bash
+pulumi config set --path 'lambdawithsecrets.envvars["envvar1"]' envvar1value
+pulumi config set --path 'lambdawithsecrets.envvars["envvar2"]' envvar2value
+```
+
+and then we add the secret settings:
+
+```bash
+pulumi config set --path 'lambdawithsecrets.secrets["secret1"]' secretvalue1
+pulumi config set --path 'lambdawithsecrets.secrets["secret2"]' secretvalue2
+```
+
+If we take a look at the stack config (in my stack named `dev` it's the `Pulumi.dev.yaml` file):
+
+```yaml
+config:
+  lambda-secrets:lambdawithsecrets:
+    envvars:
+      envvar1: envvar1value
+      envvar2: envvar2value
+    secrets:
+      secret1:
+        secure: AAABALZIAzDaE45MbdO08w/QMHK6CE4hssidMK70pyYt6hIFHpKeq0mBNic=
+      secret2:
+        secure: AAABAI+y9HpdgQ1u1rSPKGZeBkAIIbZfki84zmRrcWVltIMI5SGWZbYOyfQ=
+```
+
+there are several things I want to point out:
+
+1. The entire config item is named `lambdawithsecrets` (this is how we'll retrieve the configuration later)
+1. There are two sections, `envvars` and `secrets`. These are how I'll refer to the dictionaries later on in the code, but you can name these in any way you want.
+1. We'll use the dictionary keys (`envvar1`, `envvar2`, `secret1`, `secret2`) to refer to the variables in the Lambda Function so we know what they're named and can easily access them. Again, you can name these to be whatever values you want.
+
+## Writing our Pulumi program
+
+Next we're going to start updating our `index.ts` file. Everything from now on can be copied and pasted in order into the file, and [I'll post the entire contents at the end]({{< relref "#complete-code" >}}).
+
+As we're only going to use the Pulumi and the AWS NPM packages you can replace everything in the file with the following:
+
+```typescript
+import * as pulumi from "@pulumi/pulumi";
+import * as aws from "@pulumi/aws";
+```
+
+### A warning about accessing the configuration
+
+Before we continue, it's worth pointing out the following:
+
+{{% notes type="warning" %}}
+On `pulumi up`, secret values are decrypted and made available in plaintext at runtime. These may be read through any of the standard `pulumi.Config` getters. While it is possible to read a secret using the ordinary non-secret getters, this is almost certainly not what you want. Use the secret variants of the configuration APIs instead, since this ensures that all transitive uses of that secret are themselves also marked as secrets.
```

Comment:
```suggestion
On `pulumi up`, secret values are decrypted and made available in plaintext at runtime. These may be read through any of the standard `pulumi.Config` getters. While it is possible to read a secret using ordinary non-secret getters, this is almost certainly not what you want. Use the secret variants of the configuration APIs instead since this ensures that all transitive uses of that secret are themselves also marked as secrets.
```